### PR TITLE
nixos/gotosocial: init

### PR DESCRIPTION
**Pull request by misuzu**
###### Description of changes
https://gotosocial.org/
https://github.com/superseriousbusiness/gotosocial

Based on https://github.com/NixOS/nixpkgs/pull/202110

cc @f0x52

###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [X] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [X] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by figsoda** [comment]
@ofborg eval

File: pkgs/servers/gotosocial/default.nix [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+  ldflags = [
+    "-s"
+    "-w"
+    "-X main.Version=${version}"
+  ];
+
+  postInstall = ''
+    tar xf ${web-assets}
+    mkdir -p $out/share
+    mv web $out/share/
```
Comment (pbsds) [inline pkgs/servers/gotosocial/default.nix:42]: The folder name `web` could cause a collision in a linkfarm/lfs. I commonly see other packages use a folder name that match the package, i.e. `$out/share/gotosocial`

**Review by pbsds — CHANGES_REQUESTED** [review]
This looks great!

File: pkgs/servers/gotosocial/default.nix (hunk 1 above)
Comment (misuzu) [reply pkgs/servers/gotosocial/default.nix]: Fixed, thanks!

**Review by pbsds — APPROVED** [review]

File: pkgs/servers/gotosocial/default.nix [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+  owner = "superseriousbusiness";
+  repo = "gotosocial";
+
+  version = "0.8.1";
+  source-hash = "sha256-nNkcK9MmhL6HbiR8yE2OI/XbBATyJ8qNIUDmJwXxD6I=";
+  web-assets-hash = "sha256-PGllF0aGI7fSF0U6MIkgeGc+Oa0eEdiG0OWuHCe67Ko=";
+
+  web-assets = fetchurl {
+    url = "https://github.com/${owner}/${repo}/releases/download/v${version}/${repo}_${version}_web-assets.tar.gz";
+    sha256 = web-assets-hash;
```
Comment (SuperSandro2000) [inline pkgs/servers/gotosocial/default.nix:18]:
```suggestion
    hash = web-assets-hash;
```

File: nixos/modules/services/web-apps/gotosocial.nix [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+
+    storage-local-base-path = "/var/lib/gotosocial/storage";
+
+    db-type = "sqlite";
+    db-address = "/var/lib/gotosocial/database.sqlite";
+  };
+  gotosocial-admin = pkgs.writeShellScriptBin "gotosocial-admin" ''
+    exec systemd-run \
+      -u gotosocial-admin.service \
+      -p DynamicUser=yes \
```
Comment (SuperSandro2000) [inline nixos/modules/services/web-apps/gotosocial.nix:22]: Does dynamic user scale with lots of files? I think it is questionable to use it here

File: nixos/modules/services/web-apps/gotosocial.nix [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+    enable = lib.mkEnableOption (lib.mdDoc "ActivityPub social network server");
+
+    package = lib.mkPackageOptionMD pkgs "gotosocial" { };
+
+    openFirewall = lib.mkOption {
+      type = lib.types.bool;
+      default = false;
+      description = lib.mdDoc ''
+        Open the configured port in the firewall.
+        We highly reccommend using a reverse proxy instead.
```
Comment (SuperSandro2000) [inline nixos/modules/services/web-apps/gotosocial.nix:43]:
```suggestion
        Using a reverse proxy instead is highly reccommended.
```

File: nixos/modules/services/web-apps/gotosocial.nix [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+    };
+
+    configFile = lib.mkOption {
+      type = lib.types.path;
+      description = lib.mdDoc ''
+        Path to configuration file.
+        When set this overrides the generated configuration from module options.
+      '';
+      example = "/root/nixos/secrets/gotosocial.yml";
+    };
```
Comment (SuperSandro2000) [inline nixos/modules/services/web-apps/gotosocial.nix:63]: I don't think we should have such an option. Can't we supply an extra config file where the secrets can live?

File: nixos/modules/services/web-apps/gotosocial.nix [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+          "CAP_BLOCK_SUSPEND"
+          "CAP_WAKE_ALARM"
+          "CAP_SYS_TTY_CONFIG"
+          "CAP_MAC_ADMIN"
+          "CAP_MAC_OVERRIDE"
+          "CAP_NET_ADMIN"
+          "CAP_NET_BROADCAST"
+          "CAP_NET_RAW"
+          "CAP_SYS_ADMIN"
+          "CAP_SYS_PTRACE"
```
Comment (SuperSandro2000) [inline nixos/modules/services/web-apps/gotosocial.nix:158]: Why does the program need all of those caps?

File: nixos/modules/services/web-apps/gotosocial.md [hunk 7]
```diff
@@ -0,0 +1,64 @@
+# GoToSocial {#module-services-gotosocial}
+
+[GoToSocial](https://gotosocial.org/) is an ActivityPub social network server, written in Golang.
+
+## Service configuration {#modules-services-gotosocial-service-configuration}
+
+The following configuration sets up the PostgreSQL as database backend and binds
+GoToSocial to `127.0.0.1:31337`, expecting to be run behind a HTTP proxy on `gotosocial.example.com`.
```
Comment (SuperSandro2000) [inline nixos/modules/services/web-apps/gotosocial.md:8]:
```suggestion
GoToSocial to `127.0.0.1:3137`, expecting to be run behind a HTTP proxy on `gotosocial.example.com`.
```
Using such a high port could conflict with other things maybe, we should stick to something below 10k

File: pkgs/servers/gotosocial/default.nix (hunk 2 above)
Comment (misuzu) [reply pkgs/servers/gotosocial/default.nix]: Fixed!

File: nixos/modules/services/web-apps/gotosocial.nix (hunk 3 above)
Comment (misuzu) [reply nixos/modules/services/web-apps/gotosocial.nix]: Removed

File: nixos/modules/services/web-apps/gotosocial.nix (hunk 4 above)
Comment (misuzu) [reply nixos/modules/services/web-apps/gotosocial.nix]: Updated, thanks!

File: nixos/modules/services/web-apps/gotosocial.nix (hunk 5 above)
Comment (misuzu) [reply nixos/modules/services/web-apps/gotosocial.nix]: I've added `environmentFile` instead.

File: nixos/modules/services/web-apps/gotosocial.nix (hunk 6 above)
Comment (misuzu) [reply nixos/modules/services/web-apps/gotosocial.nix]: Removed.

File: nixos/modules/services/web-apps/gotosocial.md (hunk 7 above)
Comment (misuzu) [reply nixos/modules/services/web-apps/gotosocial.md]: Fixed!

**Comment by f0x52** [comment]
Nice work! Some notes:
- I think it would be good to explicitly mention that the `host` config cannot be changed after running once, similar to matrix-synapse
- The package definition could also build the frontend assets with something similar to https://git.pixie.town/f0x/nixos/src/branch/main/nodes/cosmos/services/gotosocial/package.nix

**Comment by misuzu** [comment]
> * I think it would be good to explicitly mention that the `host` config cannot be changed after running once, similar to matrix-synapse

Added a note, thanks!



> * The package definition could also build the frontend assets with something similar to https://git.pixie.town/f0x/nixos/src/branch/main/nodes/cosmos/services/gotosocial/package.nix

That would be very ugly, I think it's better to wait for https://github.com/NixOS/nixpkgs/pull/210814 to be merged first.

**Review by SuperSandro2000 — COMMENTED** [review]
please rebase otherwise LGTM

**Review by SuperSandro2000 — APPROVED** [review]
please do another rebase

**Pull request merged**